lbrynet/conf.py

```diff
@@ -119,276 +102,289 @@ def _convert_value(self, value):
+    'upload_log': (bool, True),
+    'use_auth_http': (bool, False),
+    'use_upnp': (bool, True),
+    'wallet': (str, LBRYUM_WALLET),
+}
 
-    def get_conf_filename(self):
-        return get_settings_file_ext(self.ensure_data_dir())
 
+class Config:
+    def __init__(self, fixed_defaults, adjustable_defaults, conf_file_settings={}, environment=None,
+                 cli_settings={}):
+        self._lbry_id = None
+        self._session_id = base58.b58encode(utils.generate_id())
 
-def update_settings_from_file(filename=None):
-    filename = filename or settings.get_conf_filename()
-    try:
-        updates = load_settings(filename)
-        log.info("Loaded settings file: %s", updates)
-        settings.update(updates)
-    except (IOError, OSError) as ex:
-        log.info('%s: Failed to update settings from %s', ex, filename)
+        self.__fixed = fixed_defaults
+        self.__adjustable = adjustable_defaults
 
+        self._init_defaults()
+        self.conf_file_settings = conf_file_settings  # from daemon_settings.yml
+        self.env_settings = self._parse_environment(environment)  # from environment variables
+        self.cli_settings = cli_settings  # from command_line flags/args
+        self.runtime_settings = {}  # set by self.set() during runtime
+        self._assert_valid_settings()
+        self._combine_settings()
 
-def get_settings_file_ext(data_dir):
-    yml_path = os.path.join(data_dir, "daemon_settings.yml")
-    json_path = os.path.join(data_dir, "daemon_settings.json")
-    if os.path.isfile(yml_path):
-        return yml_path
-    elif os.path.isfile(json_path):
-        return json_path
-    else:
-        return yml_path
+    def __repr__(self):
+        return self.combined_settings.__repr__()
 
+    def __iter__(self):
+        for k in self.combined_settings.iterkeys():
+            yield k
 
-settings_decoders = {
-    '.json': json.loads,
-    '.yml': yaml.load
-}
+    def __getitem__(self, name):
+        return self.get(name)
+
+    def __setitem__(self, name, value):
+        return self.set(name, value)
+
+    def __contains__(self, name):
+        return name in self.combined_settings
+
+    def _parse_environment(self, environment):
+        env_settings = {}
+        if environment is None:
+            environment = Env(**self.__adjustable)
+        if environment:
+            assert isinstance(environment, Env)
+            for opt in environment.original_schema:
+                env_settings[opt] = environment(opt)
+        return env_settings
+
+    def _init_defaults(self):
+        self.defaults = self.__fixed.copy()
+        for k, v in self.__adjustable.iteritems():
+            self.defaults[k] = v[1]
+
+    def _assert_valid_settings(self):
+        for s in [self.conf_file_settings, self.env_settings, self.cli_settings,
+                  self.runtime_settings]:
+            for name in s:
+                assert name in self.defaults, IndexError('%s is not a valid setting' % name)
```

Comment:
im not sure whether this is a case of bad user input. my understanding is that assertions are special because they can be turned off, so you cant rely on them for validating things like "did the user enter a number when we asked for one". but in this case, they are being used for internal validation. i can see validating the config file as user input, but not the get()/set() methods. does that make sense?